File: Python vs Java/3 - Branching/Branching.java
/**
 * Need a bit more ceremonial code here for Java, which is kind of a gift and a curse.
 * 
 * Java Version: 25.0.3
 * 
 * @author Jon Mrowczynski
 */
void main() {
    // Boolean values are "true" or "false"

    var iAmTrue = true;
    var iAmFalse = false;
    var iAmAlsoTrue = true;

    // if-if else-else branching is supported (more standard "else if" is used)
    // Java uses curly braces to associated code.
    if (iAmTrue) {
        IO.println("This executed because the if condition was True.");
    } else if (iAmAlsoTrue) {
        IO.println("Even though the else if condition is also True, since the if condition was True, this will not execute.");
    } else {
        IO.println("This would only execute if all other conditions were false.");
    }

    var iAmDependent = (iAmTrue) ? 1 : 0; // Java has conditional expressions commonly used for assignments.
    IO.println(iAmDependent);

    switchExpressionExample();
    switchStatementExample();

    IO.println();

    andExamples();
    orExamples();
    notExamples();
}

/**
 * Java jas more standard switch expression that has grown relatively powerful with recent versions.
 */
void switchExpressionExample() {
    var pattern = "Set me to another String";
    switch(pattern) {
        case "Hello":
            IO.println("Do something here if String is \"Hello\".");
            break; // An explicit break statement is required to stop potentially running more cases.
        case "World!":
            IO.println("Do something here is String is \"World!\".");
            break;
        default:
            IO.println("This is run if pattern doesn't match any other case.");
            // Don't need to explicitly break default.
    }
}

/**
 * Note that a switch statement doesn't use breaks! This is done for you behind the scenes. switch statements can also work
 * on other data types, not just Strings.
 */
void switchStatementExample() {
    var pattern = "Set me to something";
    var result = switch(pattern) {
        case "Hello" -> 1;
        case "World!" -> 2;
        default -> 3;
    };
    IO.println(result); // The value of result depends on the value of pattern!
}

/**
 * Shows logical results using the and '&&' operator. The first two boolean expressions have dead code on the right due to
 * short-circuiting. This shouldn't be confused with the bitwise and '&'.
 */
void andExamples() {
    IO.println("""
            And Examples:
            false && false: %b
            false && true: %b
            true %% false: %b
            true && true: %b
            """.formatted(false && false, false && true, true && false, true && true));
}

/**
 * Shows logical results using the or '||' operator. The last two boolean expressions have dead code on the right due to
 * short-circuiting. This shouldn't be confused with the bitwise or '|'.
 */
void orExamples() {
    IO.println("""
            Or Examples:
            false || false: %b
            false || true: %b
            true || false: %b
            true || true: %b
            """.formatted(false || false, false || true, true || false, true || true));
}

/**
 * Shows logical results using the not '!' operator. This shouldn't be confused with the bitwise not '~'.
 */
void notExamples() {
    IO.println("""
            Not Examples:
            !false: %b
            !true: %b
            """.formatted(!false, !true));
}

File: Python vs Java/3 - Branching/branching.py
"""
Author: Jon Mrowczynski
Python Version: 3.13.7
"""

# Boolean values are "True" or "False".

i_am_true = True
i_am_false = False
i_am_also_true = True

# if-elif-else branching is supported, although Python shortens "if else" to "elif".
# Also, indentation is used to associated code.

if i_am_true:
    print('This executed because the if condition was True.')
elif i_am_also_true:
    print('Even though the elif condition is also True, since the if condition was True, this will not execute.')
else:
    print('This would only execute if all other conditions were False.')

# Can use conditional expressions commonly used for assignments.
i_am_dependent = 1 if i_am_true else 0
print(i_am_dependent)

# Python does have a "switch" style mechanism (introduced in 3.10), but it doesn't seem to be widely used yet...
# However, it can help to improve the readability by cleaning up more complex if-elif-else logic branching.
pattern = "Set me to something"
match pattern:
    case 'a':
        print('Do something here if "pattern" is "a".')
    case 3:
        print('Do something here if "pattern" is 3.')
    case _:
        print('This is the catch all case that runs if "pattern" doesn\'t match any of the other cases.')

print()
# Logical operators include "and", "or", and "not".
print('False and False:', False and False)  # The expression on the right is dead code due to short-circuiting.
print('False and True:', False and True)  # The expression on the right is dead code due to short-circuiting.
print('True and False:', True and False)
print('True and True:', True and True)
print() 

print('False or False:', False or False)
print('False or True:', False or True)
print('True or False:', True or False)  # The expression on the right is dead code due to short-circuiting.
print('True or True:', True or True)  # The expression on the right is dead code due to short-circuiting.
print()

print('not False:', not False)
print('not True:', not False)
